src/Lecture3.hs

```diff
@@ -125,9 +154,12 @@ data Reward = Reward
     } deriving (Show, Eq)
 
 instance Semigroup Reward where
-
+  (Reward ag as) <> (Reward bg bs)
+    | as || bs = Reward (ag <> bg) True
+    | otherwise = mempty
```

Comment:
I expected a slightly different implementation but I guess this one works too because it's a perfectly lawful implementation 😅 It's not immediately obvious why this implementation of `<>` is associative. But the tests are passing so I guess this is fine!